src/Chainweb/Pact/Backend/DbCache.hs

```diff
@@ -0,0 +1,236 @@
+  , _ceData :: !a
+    -- ^ Cached data. This is stored in a compact region.
+  , _ceHits :: Int
+    -- ^ Count of cache hits for this entry
+  }
+
+ceTxId :: Lens' (CacheEntry a) TxId
+ceTxId = lens _ceTxId (\a b -> a { _ceTxId = b })
+
+ceHits :: Lens' (CacheEntry a) Int
+ceHits = lens _ceHits (\a b -> a { _ceHits = b })
+
+instance Eq (CacheEntry a) where
+  a == b = _ceAddy a == _ceAddy b && _ceTxId a == _ceTxId b
+
+-- | Sort by '_ceTxId' so that smaller tx ids are evicted first from the
+-- cache. '_ceAddy' is just for stable sort.
+--
+instance Ord (CacheEntry a) where
+  a `compare` b = cmp _ceTxId (cmp _ceAddy EQ)
+    where
+      cmp acc next = case acc a `compare` acc b of
+        EQ -> next
+        c -> c
+
+-- -------------------------------------------------------------------------- --
+-- DbCache
+
+data DbCache a = DbCache
+  { _dcStore :: !(HM.HashMap CacheAddress (CacheEntry a))
+  , _dcSize :: !Int
+  , _dcLimit :: !Int
+
+  -- Telemetry
+  , _dcMisses :: !Int
+  , _dcHits :: !Int
+  }
+
+makeLenses 'DbCache
+
+emptyDbCache :: Int -> DbCache a
+emptyDbCache limit = DbCache
+    { _dcStore = mempty
+    , _dcSize = 0
+    , _dcLimit = limit
+    , _dcMisses = 0
+    , _dcHits = 0
+    }
+
+isEmptyCache :: DbCache a -> Bool
+isEmptyCache = HM.null . _dcStore
+
+-- | The total size of all data in the cache.
+--
+-- Complexity: \(O(1)\)
+--
+cacheSize :: DbCache a -> Int
+cacheSize = _dcSize
+
+-- | Number of entries in the cache.
+--
+-- Complexity: \(O(n)\), linear in the number of entries
+--
+cacheCount :: DbCache a -> Int
+cacheCount = HM.size . _dcStore
+
+-- | Provide read-time caching of deserialized JSON values.
+-- If cache entry is found, mark it with the current txid and return value.
+-- If not, deserialize row data and store in cache, which triggers
+-- cache maintenance.
+--
+checkDbCache
+    :: FromJSON a
+    => Utf8
+        -- ^ Db key for data
+    -> ByteString
+        -- ^ row data that contains the encoded value
+    -> TxId
+       -- ^ Current TxId from checkpointer, used as priority for cache eviction.
+       -- Smaller values are evicted first.
+    -> DbCache a
+    -> IO (Maybe a, DbCache a)
+checkDbCache key rowdata txid = runStateT $ do
+    readCache txid addy >>= \case
+
+        -- Cache hit
+        Just !x -> do
+            modify' (dcHits +~ 1)
+            return $ Just x
+
+        -- Cache miss: decode module and insert into cache
+        --
+        -- The cache entry is stored within a compact region. The whole region is
+        -- GCed only when no pointer to any data in the region is live any more.
+        --
+        Nothing -> do
+            c <- lift $ compact (decodeStrict {- lazy, forced by compact -} rowdata)
+            case getCompact c of
+                Nothing -> return Nothing
+                Just m -> do
+                    s <- lift $ compactSize c
+                    writeCache txid addy (fromIntegral s) m
+                    modify' (dcMisses +~ 1)
```

Comment:
I'd kind of prefer `dcMisses +=' 1`, provided we define a set of lens operators based on `modify'`. But this is fine too.

Reply:
Agreed. Though I propose that for a separate PR that would introduce those operators and apply them pervasively across the code base. (Actually, I am pretty sure that we worked on that before in pact, but it seems it never got finalized).